File: Src/TextAnalysis/Lexer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Text;
using System.Threading.Tasks;
using Citron.LexicalAnalysis;
using Citron.Syntax;

namespace Citron
{
    public struct LexResult
    {
        public static LexResult Invalid { get; }
        static LexResult()
        {
            Invalid = new LexResult();
        }

        public bool HasValue { get; }
        public Token Token { get; }
        public LexerContext Context { get; }
        public LexResult(Token token, LexerContext context) { HasValue = true; Token = token; Context = context; }
    }

    public class Lexer
    {
        public Lexer()
        {
        }

        bool IsIdentifierStartLetter(BufferPosition curPos)
        {
            if (curPos.Equals('_')) return true; // only allowed among ConnectorPunctuation category

            var category = curPos.GetUnicodeCategory();

            return category == UnicodeCategory.UppercaseLetter ||
                category == UnicodeCategory.LowercaseLetter ||
                category == UnicodeCategory.TitlecaseLetter ||
                category == UnicodeCategory.ModifierLetter ||
                category == UnicodeCategory.OtherLetter ||
                category == UnicodeCategory.NonSpacingMark ||
                category == UnicodeCategory.LetterNumber ||
                category == UnicodeCategory.DecimalDigitNumber;
        }

        bool IsIdentifierLetter(BufferPosition curPos)
        {
            if (curPos.Equals('_')) return true; // only allowed among ConnectorPunctuation category

            var category = curPos.GetUnicodeCategory();

            return category == UnicodeCategory.UppercaseLetter ||
                category == UnicodeCategory.LowercaseLetter ||
                category == UnicodeCategory.TitlecaseLetter ||
                category == UnicodeCategory.ModifierLetter ||
                category == UnicodeCategory.OtherLetter ||
                category == UnicodeCategory.NonSpacingMark ||
                category == UnicodeCategory.LetterNumber ||
                category == UnicodeCategory.DecimalDigitNumber;
        }

        public LexResult LexStringMode(LexerContext context)
        {   
            var textResult = LexStringModeText(context);
            if (textResult.HasValue)
                return textResult;

            if (context.Pos.Equals('"'))
                return new LexResult(
                    Tokens.DoubleQuote,
                    context.UpdatePos(context.Pos.Next()));

            if (context.Pos.Equals('$'))
            {
                var nextPos = context.Pos.Next();

                if (nextPos.Equals('{'))
                    return new LexResult(
                        Tokens.DollarLBrace,
                        context.UpdatePos(nextPos.Next()));

                var idResult = LexIdentifier(context.UpdatePos(nextPos), false);
                if (idResult.HasValue)
                    return idResult;
            }            

            return LexResult.Invalid;
        }

        // 키워드 처리
        private static Dictionary<string, Token> keywordInfos = new Dictionary<string, Token>()
        {
            { "foreach", Tokens.Foreach },
            { "if", Tokens.If },
            { "else", Tokens.Else },
            { "for", Tokens.For },
            { "continue", Tokens.Continue },
            { "break", Tokens.Break },
            { "task", Tokens.Task },
            { "params", Tokens.Params },
            { "return", Tokens.Return },
            { "async", Tokens.Async },
            { "await", Tokens.Await },
            { "in", Tokens.In },
            { "yield", Tokens.Yield },
            { "seq", Tokens.Seq },
            { "enum", Tokens.Enum },
            { "struct", Tokens.Struct },
            { "class", Tokens.Class },
            { "is", Tokens.Is },
            { "as", Tokens.As },
            { "ref", Tokens.Ref },
            { "box", Tokens.Box },
            { "null", Tokens.Null },
            { "public", Tokens.Public },
            { "protected", Tokens.Protected },
            { "private", Tokens.Private },
            { "static", Tokens.Static },
            { "new", Tokens.New },
            { "namespace", Tokens.Namespace },
        };

        private static (string Text, Func<Token> Constructor)[] infos = new (string Text, Func<Token> Constructor)[]
        {   
            ("++", () => Tokens.PlusPlus),
            ("--", () => Tokens.MinusMinus),
            ("<=", () => Tokens.LessThanEqual),
            (">=", () => Tokens.GreaterThanEqual),
            ("=>", () => Tokens.EqualGreaterThan),
            ("==", () => Tokens.EqualEqual),
            ("!=", () => Tokens.ExclEqual),
            ("->", () => Tokens.MinusGreaterThan),


            ("@", () => Tokens.At),
            ("<", () => Tokens.LessThan),
            (">", () => Tokens.GreaterThan),
            (";", () => Tokens.SemiColon),
            (",", () => Tokens.Comma),
            ("=", () => Tokens.Equal),
            ("{", () => Tokens.LBrace),
            ("}", () => Tokens.RBrace),
            ("(", () => Tokens.LParen),
            (")", () => Tokens.RParen),
            ("[", () => Tokens.LBracket),
            ("]", () => Tokens.RBracket),


            ("+", () => Tokens.Plus),
            ("-", () => Tokens.Minus),
            ("*", () => Tokens.Star),
            ("/", () => Tokens.Slash),
            ("%", () => Tokens.Percent),
            ("!", () => Tokens.Excl),
            (".", () => Tokens.Dot),
            ("?", () => Tokens.Question),
            ("&", () => Tokens.Ampersand),

            (":", () => Tokens.Colon),
            ("`", () => Tokens.Backtick)
        };

        public LexResult LexNormalMode(LexerContext context, bool bSkipNewLine)
        {
            // 스킵처리
            var wsResult = LexWhitespace(context, bSkipNewLine);
            if (wsResult.HasValue)
                context = wsResult.Context;

            // 끝 처리
            if (context.Pos.IsReachEnd())
                return new LexResult(Tokens.EndOfFile, context);

            // 줄바꿈 문자
            var newLineResult = LexNewLine(context);
            if(newLineResult.HasValue)
                return new LexResult(newLineResult.Token, newLineResult.Context);

            // 여러개 먼저
            var intResult = LexInt(context);
            if (intResult.HasValue)
                return new LexResult(intResult.Token, intResult.Context);

            var boolResult = LexBool(context);
            if (boolResult.HasValue)
                return new LexResult(boolResult.Token, boolResult.Context);

            foreach (var info in infos)
            {
                var consumeResult = Consume(info.Text, context.Pos);
                if (consumeResult.HasValue)
                    return new LexResult(info.Constructor(), context.UpdatePos(consumeResult.Value));
            }

            if (context.Pos.Equals('"'))
                return new LexResult(
                    Tokens.DoubleQuote, 
                    context.UpdatePos(context.Pos.Next()));

            var keywordResult = LexKeyword(context);
            if (keywordResult.HasValue)
                return new LexResult(keywordResult.Token, keywordResult.Context);


            // Identifier 시도
            var idResult = LexIdentifier(context, true);
            if (idResult.HasValue)
                return new LexResult(idResult.Token, idResult.Context);

            return LexResult.Invalid;
        }

        public LexResult LexCommandMode(LexerContext context)
        {   
            var newLineResult = LexNewLine(context);
            if (newLineResult.HasValue)
                return new LexResult(Tokens.NewLine, newLineResult.Context);

            // TODO: \} 처리
            if (context.Pos.Equals('}'))
                return new LexResult(Tokens.RBrace, context.UpdatePos(context.Pos.Next()));
            
            if (context.Pos.Equals('$'))
            {                
                var nextDollarPos = context.Pos.Next();

                if (nextDollarPos.Equals('{'))
                {
                    return new LexResult(
                        Tokens.DollarLBrace,
                        context.UpdatePos(nextDollarPos.Next()));
                }

                if (!nextDollarPos.Equals('$'))
                {
                    var idResult = LexIdentifier(context.UpdatePos(nextDollarPos), false);
                    if (idResult.HasValue)
                        return idResult;
                }
            }

            var sb = new StringBuilder();

            // 나머지는 text모드
            while(true)
            {
                // 끝 도달
                if (context.Pos.IsReachEnd()) break;
                
                // NewLine문자
                if (context.Pos.Equals('\r') || context.Pos.Equals('\n')) break;

                // TODO: \} 처리
                if (context.Pos.Equals('}'))
                    break;
                
                if (context.Pos.Equals('$'))
                {
                    var nextDollarPos = context.Pos.Next();
                    if (nextDollarPos.Equals('$'))
                    {
                        sb.Append('$');
                        context = context.UpdatePos(nextDollarPos.Next());
                        continue;
                    }

                    break;
                }

                context.Pos.AppendTo(sb);
                context = context.UpdatePos(context.Pos.Next());
            }

            if (0 < sb.Length)
                return new LexResult(new TextToken(sb.ToString()), context);

            return LexResult.Invalid;
        }

        LexResult LexKeyword(LexerContext context)
        {
            var sb = new StringBuilder();
            BufferPosition curPos = context.Pos;

            while (IsIdentifierLetter(curPos))
            {
                curPos.AppendTo(sb);
                curPos = curPos.Next();
            }

            if (sb.Length == 0)
                return LexResult.Invalid;

            if (keywordInfos.TryGetValue(sb.ToString(), out var token))
                return new LexResult(token, context.UpdatePos(curPos));            

            return LexResult.Invalid;
        }

        LexResult LexIdentifier(LexerContext context, bool bAllowRawMark)
        {
            var sb = new StringBuilder();
            BufferPosition curPos = context.Pos;

            if (bAllowRawMark && curPos.Equals('@'))
            {
                curPos = curPos.Next();
            }
            else if (IsIdentifierStartLetter(curPos))
            {
                curPos.AppendTo(sb);
                curPos = curPos.Next();
            }
            else
            {
                return LexResult.Invalid;
            }

            while (IsIdentifierLetter(curPos))
            {   
                curPos.AppendTo(sb);
                curPos = curPos.Next();
            }

            if (sb.Length == 0)
                return LexResult.Invalid;

            return new LexResult(new IdentifierToken(sb.ToString()), context.UpdatePos(curPos));
        }

        BufferPosition? Consume(string text, BufferPosition pos)
        {
            foreach (var c in text)
            {
                if (!pos.Equals(c))
                    return null;

                pos = pos.Next();
            }

            return pos;
        }

        LexResult LexBool(LexerContext context)
        {
            var trueResult = Consume("true", context.Pos);
            if (trueResult.HasValue)
                return new LexResult(new BoolToken(true), context.UpdatePos(trueResult.Value));

            var falseResult = Consume("false", context.Pos);
            if (falseResult.HasValue)
                return new LexResult(new BoolToken(false), context.UpdatePos(falseResult.Value));

            return LexResult.Invalid;
        }

        internal LexResult LexInt(LexerContext context)
        {
            var sb = new StringBuilder();
            BufferPosition curPos = context.Pos;

            while (curPos.GetUnicodeCategory() == UnicodeCategory.DecimalDigitNumber)
            {   
                curPos.AppendTo(sb);
                curPos = curPos.Next();
            }

            if (sb.Length == 0)
                return LexResult.Invalid;

            return new LexResult(new IntToken(int.Parse(sb.ToString())), context.UpdatePos(curPos));
        }

        LexResult LexStringModeText(LexerContext context)
        {
            var sb = new StringBuilder();
            var curPos = context.Pos;
            while (true) // 조심
            {
                if (curPos.IsReachEnd())
                    break;

                if (curPos.Equals('"')) // "두개 처리
                {
                    var secondPos = curPos.Next();
                    if (!secondPos.Equals('"')) break;

                    sb.Append('"');
                    curPos = secondPos.Next();
                }
                else if (curPos.Equals('$')) // $ 처리
                {
                    var secondPos = curPos.Next();
                    if (!secondPos.Equals('$')) break;
                    
                    sb.Append('$');
                    curPos = secondPos.Next();
                }
                else
                {
                    curPos.AppendTo(sb);
                    curPos = curPos.Next();
                }
            }

            if (sb.Length == 0)
                return LexResult.Invalid;

            return new LexResult(new TextToken(sb.ToString()), context.UpdatePos(curPos));
        }
        
        internal LexResult LexWhitespace(LexerContext context, bool bIncludeNewLine)
        {
            LexResult? nextLineModeFailedResult = null;

            bool bUpdated = false;
            while(true)
            {
                if (context.Pos.Equals('\\'))
                {
                    nextLineModeFailedResult = bUpdated ? new LexResult(Tokens.Whitespace, context) : LexResult.Invalid;
                    context = context.UpdatePos(context.Pos.Next());
                    continue;
                }

                // 코멘트 처리
                var commentBeginPos = Consume("//", context.Pos);

                if (!commentBeginPos.HasValue)
                    commentBeginPos = Consume("#", context.Pos);

                if (commentBeginPos.HasValue)
                {
                    context = context.UpdatePos(commentBeginPos.Value);

                    while (!context.Pos.IsReachEnd() && !context.Pos.Equals('\r') && !context.Pos.Equals('\n'))
                    {
                        context = context.UpdatePos(context.Pos.Next());
                        bUpdated = true;
                    }
                    continue;
                }

                if (context.Pos.IsWhiteSpace())
                {
                    context = context.UpdatePos(context.Pos.Next());
                    bUpdated = true;
                    continue;
                }

                if (bIncludeNewLine && (context.Pos.Equals('\r') || context.Pos.Equals('\n')))
                {
                    context = context.UpdatePos(context.Pos.Next());
                    bUpdated = true;
                    continue;
                }

                if (nextLineModeFailedResult.HasValue)
                {
                    var rnPos = Consume("\r\n", context.Pos);
                    if (rnPos.HasValue)
                    {
                        nextLineModeFailedResult = null;
                        context = context.UpdatePos(rnPos.Value);
                        bUpdated = true;
                        continue;
                    }
                    else if (context.Pos.Equals('\r') || context.Pos.Equals('\n'))
                    {
                        nextLineModeFailedResult = null;
                        context = context.UpdatePos(context.Pos.Next());
                        bUpdated = true;
                        continue;
                    }
                    else
                    {
                        // \ 이전을 리턴한다
                        return nextLineModeFailedResult.Value;
                    }
                }

                break;
            }

            return bUpdated ? new LexResult(Tokens.Whitespace, context) : LexResult.Invalid;
        }

        internal LexResult LexNewLine(LexerContext context)
        {
            bool bUpdated = false;
            while (context.Pos.Equals('\r') || context.Pos.Equals('\n'))
            {
                context = context.UpdatePos(context.Pos.Next());
                bUpdated = true;
            }

            return bUpdated ? new LexResult(Tokens.NewLine, context) : LexResult.Invalid;
        }
    }
}

File: src/TextAnalysis/Lexer.cpp
#include "pch.h"
#include "Lexer.h"

#include <sstream>
#include <string_view>
#include <vector>
#include <unordered_map>

#include <Syntax/Tokens.h>
#include <cassert>
#include <utf8.h>

using namespace std;
using namespace utf8;

namespace {

using namespace Citron;

struct TokenInfo
{
    const char32_t* sequence;
    Token token;    
};

TokenInfo infos[] = {
    { U"++", PlusPlusToken() },
    { U"--", MinusMinusToken() },
    { U"<=", LessThanEqualToken() },
    { U">=", GreaterThanEqualToken() },
    { U"=>", EqualGreaterThanToken() },
    { U"==", EqualEqualToken() },
    { U"!=", ExclEqualToken() },
    { U"->", MinusGreaterThanToken() },

    { U"@", AtToken() },
    { U"<", LessThanToken() },
    { U">", GreaterThanToken() },
    { U";", SemiColonToken() },
    { U",", CommaToken() },
    { U"=", EqualToken() },
    { U"{", LBraceToken() },
    { U"}", RBraceToken() },
    { U"(", LParenToken() },
    { U")", RParenToken() },
    { U"[", LBracketToken() },
    { U"]", RBracketToken() },


    { U"+", PlusToken() },
    { U"-", MinusToken() },
    { U"*", StarToken() },
    { U"/", SlashToken() },
    { U"%", PercentToken() },
    { U"!", ExclToken() },
    { U".", DotToken() },
    { U"?", QuestionToken() },
    { U"&", AmpersandToken() },

    { U":", ColonToken() },
    { U"`", BacktickToken() },
};

// 키워드 처리
unordered_map<u32string, Token> keywordInfos{
    { U"foreach"s, ForeachToken() },
    { U"if"s, IfToken() },
    { U"else"s, ElseToken() },
    { U"for"s, ForToken() },
    { U"continue"s, ContinueToken() },
    { U"break"s, BreakToken() },
    { U"task"s, TaskToken() },
    { U"params"s, ParamsToken() },
    { U"out"s, OutToken() },
    { U"return"s, ReturnToken() },
    { U"async"s, AsyncToken() },
    { U"await"s, AwaitToken() },
    { U"in"s, InToken() },
    { U"yield"s, YieldToken() },
    { U"seq"s, SeqToken() },
    { U"enum"s, EnumToken() },
    { U"struct"s, StructToken() },
    { U"class"s, ClassToken() },
    { U"is"s, IsToken() },
    { U"as"s, AsToken() },
    { U"ref"s, RefToken() },
    { U"box"s, BoxToken() },
    { U"local"s, LocalToken() },
    { U"null"s, NullToken() },
    { U"public"s, PublicToken() },
    { U"protected"s, ProtectedToken() },
    { U"private"s, PrivateToken() },
    { U"static"s, StaticToken() },
    { U"new"s, NewToken() },
    { U"namespace"s, NamespaceToken() },
};

class BufferIterator
{
    BufferPosition pos;
    ValidBufferPosition* validPos;

public:
    BufferIterator(BufferPosition pos)
        : pos(std::move(pos))
    {
        validPos = get_if<ValidBufferPosition>(&this->pos);
    }

    BufferIterator(const BufferIterator& pos) = default;

    BufferIterator& operator=(BufferIterator&& other)
    {
        pos = std::move(other.pos);
        validPos = get_if<ValidBufferPosition>(&this->pos);
        return *this;
    }

    BufferPosition GetBufferPosition() { return pos; }

    bool IsReachedEnd()
    {
        return holds_alternative<EndBufferPosition>(pos);
    }

    bool Equals(char32_t c)
    {
        return validPos && validPos->Equals(c);
    }

    bool IsWhiteSpaceExceptLineSeparator()
    {
        return validPos && validPos->IsWhiteSpaceExceptLineSeparator();
    }

    bool IsDecimalDigitNumber()
    {
        return validPos && validPos->IsDecimalDigitNumber();
    }

    bool IsIdentifierStartLetter()
    {
        return validPos && validPos->IsIdentifierStartLetter();
    }

    bool IsIdentifierLetter()
    {
        return validPos && validPos->IsIdentifierLetter();
    }

    void AppendTo(u32string& codePoints)
    {
        assert(validPos);
        validPos->AppendTo(codePoints);
    }

    Lexer MakeLexer()
    {
        return Lexer(pos);
    }

    // return false when exception (not reach to end)
    bool Next()
    {
        if (!validPos) return false;

        auto nextPos = validPos->Next();
        if (!nextPos) return false;

        pos = std::move(*nextPos);
        validPos = get_if<ValidBufferPosition>(&pos);
        return true;
    }

    // 성공시 position이 변경
    bool Consume(const char32_t* sequence)
    {
        const char32_t* cp = sequence;
        BufferIterator inner(pos);

        while (*cp)
        {
            if (!inner.Equals(*cp)) return false;
            if (!inner.Next()) return false;
            ++cp;
        }

        pos = std::move(inner.pos);
        return true;
    }
};

optional<LexResult> ResultNextPos(Token token, BufferIterator i)
{
    if (i.IsReachedEnd())
        return LexResult{ std::move(token), i.MakeLexer() };

    if (!i.Next()) return nullopt;

    return LexResult{ std::move(token), i.MakeLexer() };
}

LexResult Result(Token token, Lexer lexer)
{
    return LexResult{ std::move(token), std::move(lexer) };
}

template<typename TInt, typename TString>
TInt ParseInt(TString& str)
{
    TInt res = 0;
    size_t size = str.size();

    for (size_t i = 0; i < size; i++)
        res = res * 10 + (str[i] - '0'); // utf8, utf16, utf32모두 같다

    return res;
}

}

namespace Citron {

Lexer::Lexer(BufferPosition pos)
    : pos(std::move(pos))
{
}

optional<LexResult> Lexer::LexStringMode()
{
    auto oTextResult = LexStringModeText();
    if (oTextResult) return oTextResult;

    BufferIterator i(pos);
    if (i.IsReachedEnd()) return nullopt;

    if (i.Equals(U'"'))
        return ResultNextPos(DoubleQuoteToken(), i);

    if (i.Equals(U'$'))
    {
        if (!i.Next()) return nullopt;

        if (i.Equals(U'{'))
            return ResultNextPos(DollarLBraceToken(), i);

        return i.MakeLexer().LexIdentifier(false);
    }

    return nullopt;
}

optional<LexResult> Lexer::LexStringModeText()
{
    u32string codePoints;

    BufferIterator i(pos);
    while (!i.IsReachedEnd())
    {
        if (i.Equals(U'"')) // "두개 처리
        {
            BufferIterator j = i;

            if (!j.Next()) return nullopt;
            if (!j.Equals(U'"')) break;

            codePoints += U'"';

            i = std::move(j);
            if (!i.Next()) return nullopt;
        }
        else if (i.Equals(U'$')) // $ 처리
        {
            BufferIterator j = i;

            if (!j.Next()) return nullopt;
            if (!j.Equals(U'$')) break;

            codePoints += U'$';

            i = std::move(j);
            if (!i.Next()) return nullopt;
        }
        else
        {
            i.AppendTo(codePoints);
            if (!i.Next()) return nullopt;
        }
    }

    if (codePoints.empty())
        return nullopt; // invalid

    std::string u8token = utf8::utf32to8(codePoints);
    return LexResult { TextToken(std::move(u8token)), i.MakeLexer() };
}

optional<LexResult> Lexer::LexNormalMode(bool bSkipNewLine)
{
    // 스킵처리
    auto oWSResult = LexWhitespace(bSkipNewLine);
    if (oWSResult)
        return oWSResult->lexer.LexNormalModeAfterSkipWhitespace();

    return LexNormalModeAfterSkipWhitespace();
}

optional<LexResult> Lexer::LexNormalModeAfterSkipWhitespace()
{
    // 끝 처리
    BufferIterator i(pos);

    if (i.IsReachedEnd())
        return LexResult{ EndOfFileToken(), i.MakeLexer() };

    // 줄바꿈 문자
    if (auto oNewLineResult = LexNewLine())
        return *oNewLineResult;

    // 여러개 먼저
    if (auto oIntResult = LexInt())
        return *oIntResult;

    if (auto oBoolResult = LexBool())
        return *oBoolResult;

    for(auto& info : infos)
    {
        if (i.Consume(info.sequence))
            return LexResult{ info.token, i.MakeLexer() };
    }

    if (i.Equals(U'"'))
        return ResultNextPos(DoubleQuoteToken(), i);

    if (auto oKeywordResult = LexKeyword())
        return *oKeywordResult;

    // Identifier 시도
    if (auto oIdResult = LexIdentifier(true))
        return *oIdResult;

    return nullopt;
}

optional<LexResult> Lexer::LexCommandMode()
{
    if (auto oNewLineResult = LexNewLine())
        return *oNewLineResult;

    BufferIterator i(pos);

    if (i.IsReachedEnd()) return nullopt;

    // TODO: \} 처리
    if (i.Equals(U'}'))
        return ResultNextPos(RBraceToken(), i);

    if (i.Equals(U'$'))
    {
        BufferIterator j = i;

        if (!j.Next()) return nullopt;
        if (j.IsReachedEnd()) return nullopt; // $로 열어놓고 끝에 도달하면, 에러

        if (j.Equals('{'))
            return ResultNextPos(DollarLBraceToken(), j);

        if (!j.Equals('$'))
        {
            if (auto oIdResult = j.MakeLexer().LexIdentifier(false))
                return *oIdResult;
        }
    }

    u32string codePoints;

    // 나머지는 text모드
    while (!i.IsReachedEnd()) // 끝 도달 전까지
    {
        // NewLine문자
        if (i.Equals(U'\r') || i.Equals(U'\n')) break;

        // TODO: \} 처리
        if (i.Equals(U'}'))
            break;

        if (i.Equals('$'))
        {
            BufferIterator j = i;

            if (!j.Next()) return nullopt;
            if (j.IsReachedEnd()) return nullopt; // $로 열어놓고 끝에 도달하면, 에러

            if (j.Equals(U'$'))
            {
                codePoints += U'$';

                i = std::move(j);
                if (!i.Next()) return nullopt;
                continue;
            }

            break;
        }

        i.AppendTo(codePoints);
        if (!i.Next()) return nullopt;
    }

    if (!codePoints.empty())
        return Result(TextToken(utf32to8(codePoints)), i.MakeLexer());
    else
        return nullopt;
}

bool Lexer::IsReachedEnd()
{
    return holds_alternative<EndBufferPosition>(pos);
}

optional<LexResult> Lexer::LexIdentifier(bool bAllowRawMark)
{
    u32string codePoints;

    BufferIterator i(pos);
    if (i.IsReachedEnd()) return nullopt;

    if (bAllowRawMark && i.Equals(U'@'))
    {
        // eat
    }
    else if (i.IsIdentifierStartLetter())
    {
        i.AppendTo(codePoints);
    }
    else
    {
        return nullopt;
    }    

    while (!i.IsReachedEnd())
    {
        // get next valid position
        if (!i.Next()) return nullopt;

        // append to code point if it is identifier letter
        if (!i.IsIdentifierLetter()) break;
        i.AppendTo(codePoints);
    }

    if (codePoints.empty()) return nullopt;
    return Result(IdentifierToken(utf32to8(codePoints)), i.MakeLexer());
}

optional<LexResult> Lexer::LexKeyword()
{
    u32string codePoints;

    BufferIterator bi(pos);

    while (bi.IsIdentifierLetter())
    {
        bi.AppendTo(codePoints);
        if (!bi.Next()) return nullopt;
    }

    if (codePoints.empty())
        return nullopt;

    auto i = keywordInfos.find(codePoints);
    if (i != keywordInfos.end())
        return Result(i->second, bi.MakeLexer());

    return nullopt;
}

optional<LexResult> Lexer::LexBool()
{
    BufferIterator i(pos);

    if (i.Consume(U"true"))
        return LexResult{ BoolToken(true), i.MakeLexer() };

    if (i.Consume(U"false"))
        return LexResult{ BoolToken(false), i.MakeLexer() };

    return nullopt;
}

optional<LexResult> Lexer::LexInt()
{
    u32string codePoints;

    BufferIterator i(pos);

    while (i.IsDecimalDigitNumber())
    {
        i.AppendTo(codePoints);
        if (!i.Next()) return nullopt;
    }

    if (codePoints.empty())
        return nullopt;

    // TODO: 에러 처리
    int number = ParseInt<int32_t>(codePoints);
    return Result(IntToken(number), i.MakeLexer());
}

optional<LexResult> Lexer::LexWhitespace(bool bIncludeNewLine)
{
    optional<LexResult> nextLineModeFailedResult = nullopt;

    BufferIterator i(pos);

    bool bUpdated = false;
    while (!i.IsReachedEnd())
    {
        if (i.Equals(U'\\'))
        {
            if (bUpdated)
                nextLineModeFailedResult = Result(WhitespaceToken(), i.MakeLexer());
            else
                nextLineModeFailedResult = nullopt;

            if (!i.Next()) return nullopt;
            continue;
        }

        // 코멘트 처리
        bool bComment = i.Consume(U"//");

        if (!bComment)
            bComment = i.Consume(U"#");

        if (bComment)
        {
            while (!i.IsReachedEnd() && !i.Equals(U'\r') && !i.Equals(U'\n'))
            {
                bUpdated = true;
                if (!i.Next()) return nullopt;
            }

            continue;
        }

        if (i.IsWhiteSpaceExceptLineSeparator())
        {
            bUpdated = true;
            if (!i.Next()) return nullopt;
            continue;
        }

        if (bIncludeNewLine && (i.Equals(U'\r') || i.Equals(U'\n')))
        {
            bUpdated = true;

            if (!i.Next()) return nullopt;
            continue;
        }

        if (nextLineModeFailedResult)
        {
            auto oRNPos = i.Consume(U"\r\n");

            if (oRNPos)
            {   
                nextLineModeFailedResult = nullopt;
                bUpdated = true;
                continue;
            }
            else if (i.Equals(U'\r') || i.Equals(U'\n'))
            {
                nextLineModeFailedResult = nullopt;                
                bUpdated = true;

                if (!i.Next()) return nullopt;
                continue;
            }
            else
            {
                // \ 이전을 리턴한다
                return *nextLineModeFailedResult;
            }
        }

        break;
    }

    if (bUpdated)
        return LexResult{ WhitespaceToken(), i.MakeLexer()};
    else
        return nullopt;
}


optional<LexResult> Lexer::LexNewLine()
{
    BufferIterator i(pos);

    bool bUpdated = false;
    while (!i.IsReachedEnd() && (i.Equals(U'\r') || i.Equals(U'\n')))
    {
        bUpdated = true;
        if (!i.Next()) return nullopt;
    }

    if (bUpdated)
        return LexResult{ NewLineToken(), i.MakeLexer() };
    else
        return nullopt;
}



}